### PR TITLE
Option map & filter implementation

### PR DESCRIPTION
**Pull request by ApexTheory**
PR for https://github.com/ergoplatform/sigma-rust/issues/360

### CODE REVIEW

File: ergotree-ir/src/mir/value.rs [hunk 1]
```diff
@@ -432,7 +432,7 @@ impl<T: TryExtractFrom<Value>> TryExtractFrom<Value> for Option<T> {
         match v {
             Value::Opt(opt) => opt.map(T::try_extract_from).transpose(),
             _ => Err(TryExtractFromError(format!(
-                "expected Context, found {:?}",
+                "expected Option, found {:?}",
```
Comment (greenhat) [inline ergotree-ir/src/mir/value.rs:435]: Thanks for the fix!

File: ergotree-ir/src/types/soption.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        name: "filter",
+        tpe: SFunc::new(
+            vec![
+                SType::SOption(SType::STypeVar(STypeVar::iv()).into()),
+                SFunc::new(
+                    vec![STypeVar::iv().into()],
+                    SType::SBoolean,
+                ).into()
+                ],
+            SType::SOption(SType::STypeVar(STypeVar::ov()).into()),
```
Comment (greenhat) [inline ergotree-ir/src/types/soption.rs:67]:
```suggestion
            SType::SOption(SType::STypeVar(STypeVar::iv()).into()),
```

File: ergotree-interpreter/src/eval/soption.rs [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+    let normalized_input_val: Option<Value> = match input_v {
+        Value::Opt(opt) => Ok(*opt),
+        _ => Err(EvalError::UnexpectedValue(format!(
+            "expected filter input to be Value::Opt, got: {0:?}",
+            input_v
+        ))),
+    }?;
+
+    if let Some(v) = normalized_input_val {
+        if let Value::Boolean(predicate_res) = predicate_call(v.clone())? {
```
Comment (greenhat) [inline ergotree-interpreter/src/eval/soption.rs:73]: I believe if `predicate_call` return something other than `Value::Boolean` we silently return `None`. I'd rather return an error in this case.

**Review by greenhat — APPROVED** [review]
Looking great overall! I appreciate the test coverage! Please see my comments.

File: ergotree-interpreter/src/eval/soption.rs (hunk 3 above)
Comment (ApexTheory) [reply ergotree-interpreter/src/eval/soption.rs:73]: So I figured that this case would be handled by `MethodCall::new` where it checks expected types according to the `SMethodDesc`.  In a unit test, I am unable to even instantiate the `MethodCall` if the predicate's `Expr.tpe()` is not `SType::SBoolean`.

In any case, I have added an explicit check for this in the eval function as well.  I switched to match statements to make it more readable.

File: ergotree-ir/src/types/soption.rs (hunk 2 above)
Comment (ApexTheory) [reply ergotree-ir/src/types/soption.rs:67]: Thanks, good catch.

File: ergotree-interpreter/src/eval/soption.rs (hunk 3 above)
Comment (greenhat) [reply ergotree-interpreter/src/eval/soption.rs]: Type check on IR node construction checks the declared type while on the evaluation we have a resulted value that is not guaranteed to have that declared type. Think of not-yet-evaluated expression as an opaque box with a sticker(type) on it. Only when it is opened (expression is evaluated) we can see the result value and its type. Due to possible errors on evaluation, compilation, serialization, etc. I prefer to "double" check that what we got is the same that what was expected (sticker) before propagating an invalid value further up the expression tree.

**Review by greenhat — APPROVED** [review]
Looking great!

File: ergotree-interpreter/src/eval/soption.rs (hunk 3 above)
Comment (ApexTheory) [reply ergotree-interpreter/src/eval/soption.rs]: Makes sense, thanks for the explanation.

**Pull request merged**